README.md
NephroWiki extension for MediaWiki
==================================

This is a MediaWiki extension for the private wiki 'NephroWiki'. It serves
to add custom JavaScript to every wiki page. This is accomplished by
[defining][2] a [resource loader][1] module which defines the path to the
custom JavaScript. I previously used the `MediaWiki:common.js` page, but
maintaining the code on this page was very cumbersome. Now I can use my
favorite programming editor along with my favorite version control software
to write and maintain the code.

This GitHub repository was set up to help anybody who wants to serve their
own JavaScript in their Wiki. It is deliberately not registered on
mediawiki.org as it serves a rather special purpose.

The extension consists of the following files:

* `NephroWiki.php` -- registers the resource loader and a parser hook that
  ensures the resource module is loaded with every page.
* `NephroWiki.body.php` -- defines the `NephroWiki` class and its
  `loadModule` method which is hooked onto the MediaWiki parser.
* `NephroWiki.js` -- contains custom JavaScript. You won't need the code in
  this file; in my Wiki, it serves to facilitate programming of online
  calculators.

If you want to use this extension on your own wiki, you may want to clone
the repository:

```bash
git clone https://bovender.github.com/NephroWiki MyOwnWiki
```

I suggest you replace all occurrences of 'NephroWiki' with something else
with a command like this (provided you are using Linux):

```bash
cd MyOwnWiki
sed -i 's/[Nn]ephro[Ww]iki/MyOwnWiki/g' *.php *.js
```

But your mileage may vary of course.


Enabling the extension for your wiki
------------------------------------

Add the following line to your `LocalSettings.php`:

```php
require_once("$IP/extensions/NephroWiki/NephroWiki.php");
```

Replace `NephroWiki` with your chosen extension name.


License
-------

This extension is licensed under the [GNU General Public License V.2][gpl].

	NephroWiki extension
	Copyright (C) 2013-2023 Daniel Kraus (github.com/bovender)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


[1]: http://www.mediawiki.org/wiki/ResourceLoader
[2]: http://www.mediawiki.org/wiki/ResourceLoader/Developing_with_ResourceLoader#Registering
[gpl]: http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html


HISTORY.md
Version History of NephroWiki Extension
=======================================

UNRELEASED
----------

- Fix fatal error due to non-static loadModule function.

1.1.2
-----
Fix URL in metadata.

1.1.1
-----
Add ability to parse checkboxes in calculator script.

1.0.0
-----

Initial public release.


NephroWiki.js
/* Nephrowiki.js
 * (c) 2013 Daniel Kraus (http://bovender.users.sf.net)
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * ------------------------------------------------------------
 *
 * This file holds the custom scripts that are used for online
 * calculators and such. 
 */


/* ***********************************************************************
   Helper functions for calculator algorithms
   ***********************************************************************
*/

// Define a class to obtain a private name space.
function NephroWiki() {}

// Retrieve the name of the currently checked radio button of a given
// group. Returns nothing if no button is checked.
NephroWiki.getRadioValue = function(groupName) {
	return $('input[name='+groupName+']:checked').val();
}

// Parse an input field and return its value
// Automatically converts ',' to decimal points and
// respects the field's min/max values.
NephroWiki.parseInputValue = function(formID, fieldName) {
	// Get the current input field as a jQuery object
	var input = $('#'+formID+' [name='+fieldName+']');

	// Locate the associated label
	var label = input.prevUntil('input', 'span:first');
	var labelText = label.text().replace(/[:=]$/, '');

	// Get the min/max parameters and the value
	var min = input.data('min');
	var max = input.data('max');

	if ( input.data('date') == undefined ) {
		var val = parseFloat(input.val().replace(',', '.'));
		if (!val) {
			throw 'Bitte ' + labelText + ' eingeben.';
		}
	} else {
		// The value of a date field is generated differently.
		var val = input.val();
		// Check if the input text matches the common German date format
		// (with optional leading zeros and 2-digit or 4-digit year).
		var date = /(\d{1,2})\.(\d{1,2})\.(\d{4}|\d{2})/.exec(val);
		if ( date ) {
			// Deal with two-digit years: If it is < 80, assume it is in the
			// current (21st) century. This will cause a Y3K problem
			// sometime, but I will let others fix it then...
			if ( date[3] < 100 ) {
				var century;
				( date[3] < 80 ) ? century = '20' : century = '19';
				date[3] = century + date[3];
			}
			// The Date constructor expects the month to be between 0 and
			// 11, so we must subtract 1 from the actual month number.
			val = new Date(date[3], date[2]-1, date[1]);
		} else {
			throw 'Bitte ' + labelText + ' in der Form "T.M.JJ" ' +
				'oder "TT.MM.JJJJ" eingeben.';
		}
	}

	// If the field has a boolean "percent" data attribute,
	// and the user has entered a decimal number between 0 and 1,
	// multiply the value by 100.
	if ((input.data('percent')) && (Math.abs(val) <= 1)) {
		val *= 100;
	};

	// If the entered value is outside the acceptable
	// range, throw an error.
	if (val < min) {
		throw labelText + ' darf nicht kleiner als ' + min + ' sein.';
	} else if (val > max) {
		throw labelText + ' darf nicht größer als ' + max + ' sein.';
	};
	return val;
}

// Helper function that parses a form's input fields and creates a hash from
// them, which it will then return.
// Cf.  http://andrewdupont.net/2006/05/18/javascript-associative-arrays-considered-harmful
// for more on hashes aka 'associative arrays' in JavaScript.
NephroWiki.parseForm = function(formID) {
	var params = new Object();
	$('#'+formID).find('input[type=text]').each(function() {
		var fieldName = $(this).attr('name');
		params[fieldName] = NephroWiki.parseInputValue(formID, fieldName);
	});
	$('#'+formID).find('input:checked').each(function() {
		var fieldName = $(this).attr('name');
		params[fieldName] = $(this).attr('value');
	});
	// TODO: parse other fields as well (e.g., selection lists).
	return params;
}


// The calculate method is a wrapper for the various functions that perform
// the actual calculations. It takes care of form evaluation and error
// handling, allowing us to keep the actual calculator functions DRY.
NephroWiki.calculate = function(formID) {
	var form = $('#'+formID);
	try {
		// Parse the form and generate a hash.
		var params = this.parseForm(formID);
		// Call the calculation function by the name of the form ID.
		// See: http://stackoverflow.com/questions/912596
		var result = window[formID](params);
		// Output the result.
		form.find('.result').html(result);
	} catch (e) {
		form.find('.result').html('<div class="error">' + e + '</div>');
	} finally {
		result = undefined;
		params = undefined;
	}
}



/* ***********************************************************************
   Initialization method to set up a page's calculator forms.
   ***********************************************************************
*/
$(document).ready(function() { 

	// We need several wrapper functions the create closures for
	// the callback functions. This is required to be able to dynamically
	// set up the callbacks for any form.
	// Generate callback closure for the slide method.
	function slideCallBack(formID, inputField) {
		return function(event, ui) {
			inputField.val(ui.value);
			// Call the function indicated by the form's ID.
			NephroWiki.calculate(formID);
		}
	}

	// Generate callback closure for the click method.
	function clickCallBack(formID) {
		return function() {
			// Call the function indicated by the form's ID.
			NephroWiki.calculate(formID);
		}
	}

	// Generate callback closures for the keyup methods.
	// We abandon the DRY principle here so that we don't have to
	// put the detection algorithm for associated sliders into
	// the callbacks.
	function keyUpCallBack(formID) {
		return function(event) {
			// Call the function indicated by the form's ID.
			NephroWiki.calculate(formID);
		}
	}
	function keyUpCallBackWithSlider(formID, inputField, slider) {
		return function(event) {
			// Update the associated slider
			slider.slider('value', inputField.val());
			// Call the function indicated by the form's ID.
			NephroWiki.calculate(formID);
		}
	}

	// Actual body of the function:
	// Iterate over all calculator forms on the page and attach the
	// appropriate event handlers as needed.
	$('.nwCalc').each(function() {
		// Get the name of the current form; we need it in the callback
		// closures.
		var formID = $(this).attr('id');
		if (!formID) {
			throw "Missing ID on form.nwCalc!"
		}

		// Check if the form tells us how wide the sliders should be.
		// If the data attribute is not present, use a sensible default.
		var sliderWidth = $(this).data('slider-width');
		if (!sliderWidth) {
			sliderWidth = 300; // pixels
		}

		// Iterate through all the slider divs inside this form
		$(this).find('.nwSlider').each(function() {
			// Get the input field that belongs to this slider.
			// The name of the input field that is associated with a slider
			// div is stored in the slider div's data-field attribute.
			var slider = $(this);
			var fieldName = slider.data('field');
			if (!fieldName) {
				throw "Missing data-field on nwSlider."
			}
			var inputField = $('#'+formID+' [name='+fieldName+']');
			if (!inputField.length) {
				throw "Missing input field associated with nwSlider."
			}

			// Retrieve the slider's parameters from the associated
			// input element which is indicated in the slider's data-field
			var minVal = inputField.data('min');
			var maxVal = inputField.data('max');
			var defaultVal = inputField.data('default');

			// The stepper value is taken from the slider's data
			// attribute (it would not make sense to define it in the
			// input field's attributes).
			var stepVal = slider.data('step');

			// Make the div a jQueryUI slider and attach the required callback
			// functions.
			slider.slider({
				value: defaultVal, min: minVal, max: maxVal, step: stepVal, 
				slide: slideCallBack(formID, inputField)
			});

			// Put the default value into the input field.
			if (inputField.length) {
				inputField.val(defaultVal);
			}

			// Set the slider's width
			slider.css('width', sliderWidth + 'px');
		});

		// Iterate through all the text input fields (which may or may not
		// belong to a slider div).
		$(this).find('input[type=text]').each(function() {
			var inputField = $(this);
			var slider = $('#'+formID+
				' .nwSlider[data-field='+inputField.attr('name')+']');
			// If the input field has an associated slider, the slider
			// object will have a length greater than zero.
			if (slider.length) {
				inputField.keyup(keyUpCallBackWithSlider(formID, inputField,
						slider));
			} else {
				inputField.keyup(keyUpCallBack(formID));
			}
		});

		$(this).find('input[type=text][data-date]').each(function() {
			$(this).datepicker({
				shortYearCutoff: "+1",
				showOn: "button",
				onSelect: clickCallBack(formID)
			});
			$(this).css('text-align', 'left');
		});

		// Iterate through all radio buttons on this form.
		$(this).find('input[type=radio], input[type=checkbox]').each(function() {
			$(this).click(clickCallBack(formID));
		});

		// Iterate through all span elements and make them equally wide
		// This requires the spans to have display: inline-block
		var spanWidth = 0;
		$(this).find('span').each(function() {
			if ($(this).width() > spanWidth) {
				spanWidth = $(this).width();
			}
		});
		$(this).find('span').each(function() {
			$(this).width(spanWidth);
		});

		NephroWiki.calculate(formID);
	});
});

/* vim: set tw=76 fo=tqn: */


NephroWiki.body.php
<?php
# NephroWiki extension for MediaWiki
# (c) 2013 Daniel Kraus (http://bovender.users.sf.net)
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 2
# as published by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# --------------------------------------------------------------------
#
# This file contains the NephroWiki class which defines the loadModule
# method that is hooked on to OutputPageParserOutput. It serves to add
# the resource loader module which loads NephroWiki's own JavaScript.

if ( !defined( 'MEDIAWIKI' ) ) {
	die( 'Not an entry point.' );
}

class NephroWiki {
	static function loadModule($out, $parseroutput) {
		$out->addModules('ext.nephrowiki');
		return true;
	}
}


NephroWiki.Plasma.js
/* Nephrowiki.Plasma.js
 * (c) 2013 Daniel Kraus (http://bovender.users.sf.net)
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * ------------------------------------------------------------
 *
 * Plasma volume calculator for therapeutic plasma exchange (TPE).
 * Wiki template name: Rechner:Plasma
 *
 * Version history:
 * 3.0  04.08.2013  Major refactoring, put code into NephroWiki extension.
 * 2.0  10.01.2012  Überarbeitung, jetzt mit jQuery und Slider-Elementen.
 */


function nwPlasma() {
	// Text-Eingabefelder auslesen und in Gleitkommazahlen umwandeln.
	try {
		with (document.PPform) {
			var bw = parseFloat(PPbw.value.replace(',', '.'));
			var hk = parseFloat(PPhk.value.replace(',', '.'));
			var pp = parseFloat(PPpp.value.replace(',', '.'));
		};
		
		// Überprüfung der Benutzereingaben:
		// Wenn noch nicht alle Felder ausgefüllt sind, wird gar nichts
		// angezeigt; bei unplausiblen Werten wird eine Warnmeldung
		// angezeigt.
		// Der Fehler "&nbsp;" wird generiert, damit der Status-Absatz
		// immer angezeigt wird.
		if ((isNaN(bw)) || (isNaN(hk)) || (isNaN(pp)))  { throw "&nbsp;"; }
			
		if ((bw < 30)  || (bw > 200)) { throw "Körpergewicht muß zwischen 30 und 200 liegen!" };
		if ((pp < 0.5) || (pp > 5))   { throw "Plasmavolumen-Faktor muß zwischen 0.5 und 5 liegen!" };
		
		// Der Hämatokrit ist möglicherweise als Decimalzahl angegeben,
		// so daß der Wert jetzt in Prozent konvertiert wird.
		if ((hk > 0) && (hk < 1)) {
			hk *= 100; 
			$('[name="PPhk"]').val(hk);
			$('#PPsliderhk').slider('value', hk);
		}
		if ((hk <  10) || (hk > 70)) { throw "Hk muß zwischen 10 und 70 liegen!" };

		
		// Berechnen:
		plasmaVol = 0.07 * bw * (1-hk/100);
		totalVol = plasmaVol * pp;
		
		if (getCheckedValue(document.PPform.PPha) == "ffponly") {
		  var numFFPs = totalVol / 0.2;         // 1 FFP = 0.2 Ltr
		  var s = '<span class="ergebnis">' +
			numFFPs.toFixed(0).replace(".", ",") + ' FFPs</span>';
		} else {
		  var numFFPs = totalVol / 0.2 * 0.5;
		  var albVol  = totalVol / 2;
		  var s = '<span class="ergebnis">' +
			numFFPs.toFixed(0).replace(".", ",") + " FFPs</span> und " +
			'<span class="ergebnis">' +
			albVol.toFixed(1).replace(".", ",") + " Liter Humanalbumin 5 %</span>";
		};
					
		// Ausgeben
		$('p#PPresult').html('Plasmavolumen: <span class="ergebnis">' +
			plasmaVol.toFixed(1).replace(".", ",") + ' Liter</span><br />' +
			'Behandlungsvolumen: <span class="ergebnis">' +
			totalVol.toFixed(1).replace(".", ",") + ' Liter ' +
			'(' + s + ')</span>');
	} catch (e) {
		$('p#PPresult').html(e);
		// console.log(e); // Console ist undefiniert beim IE!
	}
	
};




NephroWiki.php
<?php
# NephroWiki extension for MediaWiki
# (c) 2013 Daniel Kraus (http://bovender.users.sf.net)
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 2
# as published by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

if ( !defined( 'MEDIAWIKI' ) ) {
	die( 'Not an entry point.' );
}

$wgExtensionCredits['parserhook'][] = array(
	'path'           => __FILE__,
	'name'           => 'NephroWiki',
	'url'            => 'https://github.com/bovender/NephroWiki',
	'author'         => '[https://www.mediawiki.org/wiki/User:Bovender Daniel Kraus]', 
	'version'        => '1.1.2',
	'description'    => 'Lädt NephroWiki-eigenes JavaScript'
);

$wgAutoloadClasses['NephroWiki'] = dirname( __FILE__ ) . '/NephroWiki.body.php';
$wgHooks['OutputPageParserOutput'][] = 'NephroWiki::loadModule';

# Register this extension with MediaWiki's resource loader 
# to be able load NephroWiki's JavaScript.
$wgResourceModules['ext.nephrowiki'] = array(
	'scripts' => 'NephroWiki.js',
	'dependencies' => array(
		'jquery.ui.slider',
		'jquery.ui.datepicker',
	),
	'localBasePath' => __DIR__,
	'remoteExtPath' => 'NephroWiki'
);
